### PR TITLE
pile customization

### PR DESCRIPTION
**Pull request by ArnoldSmith86**
This supersedes #852 after a discussion and collaboration with @RaphaelAlvez. It adds two properties (one to cards, one to piles):

`onPileCreation` - Whenever this card forms a new pile, all properties from `onPileCreation` will be copied to the pile. A pile will only form if both cards have the same value for `onPileCreation`.

`handleCSS` - The CSS is added to the handle of the pile allowing arbitrary customization.

This PR also introduces css variables for the handle CSS

Setting --phPosition and --phSize should be enough for 90% of the customizations needed.

--phPosition: -15px;
--phSize: 40px;
 
the follow CSS variables are created based on --phPosition and --phSize. For further customization you can set them directly
--phX: var(--phPosition);
--phY: var(--phPosition);
--phHeight: var(--phSize);
--phWidth: var(--phSize);
--FontSize: calc(var(--phSize)/2);

  --phHeight, --phWidth and --FontSize are used in "small" class. this class divides the seize by 2 and is triggered when the cards are smaller then 50px.
  
  --phX and --phY are used in relation to the corner the handle is on

Adding the property "text" within the `onPileCreation` property will override the widget count and put the specified text on the pile.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3905/pr-test (or any other room on that server)

**Comment by RaphaelAlvez** [comment]
Maybe that's a good opportunity to remove that pilesWith code...

**Comment by rogerl50** [comment]
I have some questions: 
1) where does the red handle on the pile containing card 63 come from? Both cards in that pile are associated with `Deck`, whose `cardDefaults` specified an `onPileCreation` with a green handle.
2) When I hover over that pile, I see only one card listed in the list of widgets at the top. Why is that? It doesn't happen with the pile to its left.

**Comment by rogerl50** [comment]
> 1. We played around a bit and the holders in the top left use `onEnter` to set `onPileCreation` on the cards themselves.
> 
>     2. No idea. I see two cards but no pile.
> 
>     3. No idea. I see two cards but no pile.
> 
> 
> I'm pretty sure 2 and 3 should be unrelated to this PR but it does seem strange.

So none of the holders show up for me in the widget list when I hover over them. Makes it kinda hard to understand what's happening here.

**Comment by ArnoldSmith86** [comment]
None of the _holders_? Or _piles_?

**Comment by rogerl50** [comment]
None of the *holders*. I can see them in edit mode, and click on them, but when I hover over them in the JSON editor, they don't show up in the list.

**Comment by RaphaelAlvez** [comment]
> None of the _holders_. I can see them in edit mode, and click on them, but when I hover over them in the JSON editor, they don't show up in the list. (This is a different problem from what I said before.)

yeah I'm having the same problem

**Comment by ArnoldSmith86** [comment]
Oh my god. Found the issue. Totally unrelated though. Remove the background widgets for now.

**Comment by RaphaelAlvez** [comment]
> Oh my god. Found the issue. Totally unrelated though. Remove the background widgets for now.

Do you mean remove the basic wigets? not sure that works...

**Comment by ArnoldSmith86** [comment]
It kinda does. When I removed `F4`, I actually removed its widget from the list instead of shifting the hotkeys. PR is nearly ready.

**Comment by RaphaelAlvez** [comment]
> It kinda does. When I removed `F4`, I actually removed its widget from the list instead of shifting the hotkeys. PR is nearly ready.

Oh I see. It was about f4... I was thinking it was about to many F keys

it works

**Comment by rogerl50** [comment]
So just to make sure I understand: `onPileCreation` is a card property just like any other, and it gets its value through the usual collection of possible locations (`cardDefaults` or `cardTypes`, or the card itself). When two cards attempt to form a pile, if they have the same `onPileCreation` property they will form a pile, and any properties in that property are copied to the pile. If they do not, no pile will be formed. Right?

So again a couple of questions:
1) what does "the same `onPileCreation` property" mean?
2) can `onPileCreation` contain any set of properties, or just the `handleCSS` property? I'm not sure what other properties would have any effect, but still.

**Comment by ArnoldSmith86** [comment]
Right.

1. JSON.stringify() must equal.
2. Any. `clickRoutine` is a good example. And now we could add other properties to piles.

**Comment by 96LawDawg** [comment]
> 2. can `onPileCreation` contain any set of properties, or just the `handleCSS` property? I'm not sure what other properties would have any effect, but still.

In a little while I'm going to be testing some things.  Grids, changeRoutine, maybe a few other ideas.  Just to see what works and what might be nice to have.  But `clickRoutine` definitely works.

**Comment by ArnoldSmith86** [comment]
`movable`, `clickable`, `fixedParent`, `linkedToSeat`. There are a few interesting properties inherited from `Widget`.

**Comment by rogerl50** [comment]
I'm feeling super-dim today...why, if I put a single card into the "red" holder in the current pr-test room, and then drop it onto the green pile containing 58, it forms a pile, but doesn't if I drop it onto either of the other green piles?

**Comment by rogerl50** [comment]
> > 2. can `onPileCreation` contain any set of properties, or just the `handleCSS` property? I'm not sure what other properties would have any effect, but still.
> 
> In a little while I'm going to be testing some things. Grids, changeRoutine, maybe a few other ideas. Just to see what works and what might be nice to have. But `clickRoutine` definitely works.

Can you test in a different room than pr-test?

**Comment by RaphaelAlvez** [comment]
> I'm feeling super-dim today...why, if I put a single card into the "red" holder in the current pr-test room, and then drop it onto the green pile containing 58, it forms a pile, but doesn't if I drop it onto either of the other green piles?

if you drop it into the holder the holder should chang the onPileCreation of the card

**Comment by rogerl50** [comment]
> > I'm feeling super-dim today...why, if I put a single card into the "red" holder in the current pr-test room, and then drop it onto the green pile containing 58, it forms a pile, but doesn't if I drop it onto either of the other green piles?
> 
> if you drop it into the holder the holder should chang the onPileCreation of the card

And it does. When it's in the holder the onPileCreation specifies color:red. So why should it drop onto any of the three green piles?

**Comment by RaphaelAlvez** [comment]
> > > I'm feeling super-dim today...why, if I put a single card into the "red" holder in the current pr-test room, and then drop it onto the green pile containing 58, it forms a pile, but doesn't if I drop it onto either of the other green piles?
> > 
> > 
> > if you drop it into the holder the holder should chang the onPileCreation of the card
> 
> And it does. When it's in the holder the onPileCreation specifies color:red. So why should it drop onto any of the three green piles?

sorry my explanation was wrong. Arnold's code in the right most holder changes the pile directly so it doesn't effect the cards. the piles themselves are not checked. just the cards

**Comment by rogerl50** [comment]
So why do cards from the red and green holders form piles when dropped onto the other holder?

**Comment by RaphaelAlvez** [comment]
> So why do cards from the red and green holders form piles when dropped onto the other holder?

Took me some time to understand it. That holder has inherit from. it is inheriting onEnter from the other holder

I'll change that

**Comment by 96LawDawg** [comment]
All the usual suspects work fine: changeRoutine, grid, scale, rotation.  I didn't test everything.

Things that don't work: faces, text, image.  Faces is not a big deal.  It is overridden by the default thing that happens for piles (options menu).  That behavior could easily be controlled through a clickRoutine if needed.  I'm not sure about text and image.  In theory it would be nice to change the text to something else like an x or a question mark.  Or maybe to make it be an image symbol.  I don't have a real use case; just thinking of possibilities.  But changing the text is pretty fundamental to what a pile is: a count of widgets.  May not be easy or desirable to be able to override the text property.

Those are my comments.  I definitely don't see any problems with what we have right now and I will approve it after others have had a chance to comment more.

**Comment by ArnoldSmith86** [comment]
You could hide the text with `font-size: 0` or something like that and `background-image` should work to put anything there.

But yeah, maybe we should add a dedicated property to override the text.

**Comment by 96LawDawg** [comment]
I didn't think of trying `background-image`.  So I did.  It doesn't work either.  I added "background-image:URL(/assets/-1721336118_941)" to the handleCSS and nothing happened.  Just for kicks, I also tried adding it to just "css" in the `onPileCreation` property and of course that didn't work either.

**Comment by rogerl50** [comment]
So the pr-test room relies on the fact that the `onEnter` code executes *before* the pile-forming code. Perhaps that should be obvious, but it wasn't to me. Is `onEnter` the first thing that happens when a widget enters a holder?

**Comment by 96LawDawg** [comment]
Just documenting what @ArnoldSmith86 just fixed in my test room.  "background-color:transparent; background-image:url('/assets/-1721336118_941'); background-size: contain; background-repeat: no-repeat;background-position: center;" does work.  Just have to magic the css a bit to get a large image to show in a small space.

**Comment by 96LawDawg** [comment]
Text now works.

**Review by 96LawDawg — APPROVED** [review]
So it's only been a few hours, but I feel like this works great.  I've tested most of the significant things people might try to do.  Approved based on testing.

File: client/css/widgets/pile.css
Comment (rogerl50) [inline client/css/widgets/pile.css:34]: I'd prefer it if these were separate variables defaulted to these values. This gives the user more control over the size of the small handle. People with poor eyesight have trouble seeing the small handles :)

**Review by rogerl50 — CHANGES_REQUESTED** [review]

File: client/css/widgets/pile.css
Comment (RaphaelAlvez) [reply client/css/widgets/pile.css:34]: hum you could just set you handle size to be 80. this is only triggered for cards that's are under 50 so you would know when to do that change.
also It would solve your problem directly.

this class is only used to avoid having to change the pile size manually if you set a card size to be almost the same size of the handle.

I'm still open to changes just point out why I don't think this is a solution to your problem

Comment (rogerl50) [reply client/css/widgets/pile.css:34]: OK, that makes sense. If a card is shrunk (using `enlarge`), what will happen to the size of the handle?

Comment (RaphaelAlvez) [reply client/css/widgets/pile.css:34]: neither scale or enlarge seen to have any effect on this.

Don't forget to solve the request if you agree with me.

**Review by RaphaelAlvez — APPROVED** [review]
Looks goot to me. This approval doesn't mean much since I was involved but I just took another look and everything looks ok

File: client/js/widgets/pile.js
Comment (RaphaelAlvez) [inline client/js/widgets/pile.js:34]: Actually wouldn't it be better to have a this.set("text",this.childCount) here?

**Review by RaphaelAlvez — CHANGES_REQUESTED** [review]

File: client/js/widgets/pile.js
Comment (RaphaelAlvez) [reply client/js/widgets/pile.js:34]: No. it breaks the whole point of setting a text value

**Review by RaphaelAlvez — APPROVED** [review]

**Review by rogerl50 — APPROVED** [review]

**Comment by ArnoldSmith86** [comment]
@bjalder26 and @robartsd, a review would be nice since this is a pretty big step. Do you agree with the concept?

**Review by bjalder26 — APPROVED** [review]
IMHO this is pretty awesome.

**Comment by 96LawDawg** [comment]
Okay, this is a very minor thing and not worth holding the PR for.  If the pile handle is moved to a non-standard location, for example, the bottom left, then the automatic handle repositioning that occurs near the edge of the screen does not work properly.  The room still thinks the pile should be in its normal place.

**Comment by 96LawDawg** [comment]
Draft piles tutorial in http://212.47.248.129:3905/Tutorial.

**Comment by RaphaelAlvez** [comment]
> Okay, this is a very minor thing and not worth holding the PR for. If the pile handle is moved to a non-standard location, for example, the bottom left, then the automatic handle repositioning that occurs near the edge of the screen does not work properly. The room still thinks the pile should be in its normal place.

I'll open an issue for that. I'll add properties that do that just change the reference in JS.

that was not on the table before because we didn't have onPileCreation but adding pile properties is doable now

**Comment by bjalder26** [comment]
Thought I might provide a rough draft for the wiki.  Of course, feel free to change anything.

| Property  | Default value       | Description           |
| --------  | ------------------- | --------------------- |
| onPileCreation| {}                  | When a card forms a pile the set of properties in `onPileCreation` will be applied to the pile in order to customize the pile handle.  Only cards with the same value for `onPileCreation` will be able to form piles. This is a simple example: "onPileCreation": { "handleCSS": "--phPosition: -15px; --phSize: 40px;"}, this is a more complicated example:   "onPileCreation": { "handleCSS": "--phX: -15px; --phY: -10px; --phHeight: 50px; --phWidth: 40px; --FontSize: 20px;", "text": "pile"}. <br /><br /> `--phX` and `--phY` control the X and Y offset of the pile handle from the corner it is on, and they both default to the value set for `--phPosition`, which defaults to -15px.  `--phHeight` and `--phWidth` control the height and width of the pile handle and both default to the value set for `--phSize`, which defaults to 40px.  `FontSize` controls the size of the font of the pile handle, and defaults to half the `--phSize`.  `text` will replace the pile's count with the value provided.  |

| Property  | Default value       | Description           |
| --------  | ------------------- | --------------------- |
| handleCSS | ''                  | Custom css for the pile's handle.  See `onPileCreation` in Card for more information. |

**Comment by ArnoldSmith86** [comment]
> Okay, this is a very minor thing and not worth holding the PR for. If the pile handle is moved to a non-standard location, for example, the bottom left, then the automatic handle repositioning that occurs near the edge of the screen does not work properly. The room still thinks the pile should be in its normal place.

I think that's a good reason to hold the PR. It should be easier to define the default corner and size. Those are pretty much the main properties of a pile handle. I'll update the PR today.

**Comment by ArnoldSmith86** [comment]
Alright. Did some more stuff:

I renamed `text` to `handleText`. I'm not sure about this. It breaks compatibility with the `LABEL` command but it feels weird to have all the handle properties start with `handle` except `text`.


`handleSize` - default `"auto"` - Pixel size of the handle (both width and height). If set to `"auto"`, it will use `40` but switch to `20` when the pile is less than 50px in any dimension.

`handleOffset` - default `15` - The handle is shifted this many pixels out of the pile.

`handlePosition` - default `"top right"` - The default position of the handle on the pile (`"top"`/`"middle"`/`"bottom"` vertically and `"left"`/`"center"`/`"right"` horizontally). Unless it's centered, the position will still flip at the border of the room. You can set it to `"static"` to prevent any automatic positioning at the room borders if you do fancy stuff with `"handleCSS"`.

**Comment by 96LawDawg** [comment]
Tested the changes.  They all work great.  handlePosition, handleOffset, and handleSize are much easier and more intuitive than the ph stuff that was in the css.  I tested the handle repositioning at the room border and it works also.  The only thing I don't like is handleText.  It is consistent within piles, but not consistent with every other widget.  That combined with making it not work with LABEL make me want to switch it back to just text.

**Comment by 96LawDawg** [comment]
The only thing that I'm confused about is `static.`  If I set `"handlePosition": "static",` then it works, but the handle is in the top left instead of top right.  And I can't add static to any other position.  For example: `"handlePosition": "static bottom left"` or `"handlePosition": "bottom left static"` do not make the handle static.

**Comment by ArnoldSmith86** [comment]
Yes, static is just `"static"`. The idea is that it never touches your css. For example if you make the handle a bag texture that covers (nearly) the entire cards, you never want the engine to apply different classes based on the pile position.

**Comment by rogerl50** [comment]
@bjalder26 Could you try to update the wiki draft for the latest changes?

**Comment by 96LawDawg** [comment]
> Yes, static is just `"static"`. The idea is that it never touches your css. For example if you make the handle a bag texture that covers (nearly) the entire cards, you never want the engine to apply different classes based on the pile position.

But shouldn't the handle appear in the normal place in the top right?

**Comment by 96LawDawg** [comment]
> @bjalder26 Could you try to update the wiki draft for the latest changes?

Done.

**Comment by 96LawDawg** [comment]
> Alright. Did some more stuff:

So `--FontSize` remains.  By the way, it is strange the font uses a capital letter rather than lowercase as the normal convention.

It looks like `phX` and all those other things still exist. Do they need to be documented in the wiki or is covering the new properties good enough.

**Comment by ArnoldSmith86** [comment]
The "normal" place is top left. "right" is a special added class. 🤷‍♂️

**Comment by 96LawDawg** [comment]
> The "normal" place is top left. "right" is a special added class. 🤷‍♂️

Right looks normal to me 🤷‍♂️
Screenshot 2021-11-26 134954

**Comment by ArnoldSmith86** [comment]
Renamed it back.

I know what you mean about top right being the default. At this point it would make the code more complicated to have `"static"` be in the top right and I actually think top left is better starting point for things where you would use `"static"`.

So I would prefer to keep it like that.

**Comment by 96LawDawg** [comment]
> I know what you mean about top right being the default. At this point it would make the code more complicated to have `"static"` be in the top right and I actually think top left is better starting point for things where you would use `"static"`.
> 
> So I would prefer to keep it like that.

As long as there is a reason and it wasn't just a typo or something.  Fine by me.

**Comment by 96LawDawg** [comment]
One other thing.  Again, not a big deal.  Is it possible to get access to the automatic count/number that shows up in a pile handle if you don't alter the text property?  I was thinking it would be cool to put an example in the tutorial of the pile handle changing colors as it gets bigger, like going from green for a few cards to red when there is a lot.  I need access to the property for a changeRoutine.  Or at least I think I do unless someone has another way.

**Comment by ArnoldSmith86** [comment]
It's not a property. It's just the count of children. `enterRoutine` and `leaveRoutine` might work and then `COUNT` your children.

**Comment by 96LawDawg** [comment]
Thanks.  That worked.  Didn't think of enter; was too focused on change.

**Comment by ArnoldSmith86** [comment]
Nobody re-approved after the new handle properties. Is it OK?

**Review by 96LawDawg — APPROVED** [review]
I approve it again based on testing.

**Review by rogerl50 — APPROVED** [review]
Approved through test and a quick code read. Is the draft wiki that was posted a while back up to date?

**Comment by 96LawDawg** [comment]
Yes.  It reflects the latest changes.  Still one outstanding question and one observation (from an earlier post of mine):

So `--FontSize` remains. By the way, it is strange the font uses a capital letter rather than lowercase as the normal convention.

It looks like `phX` and all those other things still exist. Do they need to be documented in the wiki or is covering the new properties good enough?

**Comment by ArnoldSmith86** [comment]
Hmm, they don't _really_ serve a purpose anymore. @RaphaelAlvez, do you want to keep the variables?

**Comment by RaphaelAlvez** [comment]
> Hmm, they don't _really_ serve a purpose anymore. @RaphaelAlvez, do you want to keep the variables?

Oh I agree. Let's remove it. This should stay like the spinners css

**Comment by RaphaelAlvez** [comment]
Oh @ArnoldSmith86 this is actually a problem....

We now are limiting the size and position to be equal in both coodinates

**Comment by ArnoldSmith86** [comment]
> Oh @ArnoldSmith86 this is actually a problem....
> 
> We now are limiting the size and position to be equal in both coodinates

Just override `handleCSS`.

**Comment by RaphaelAlvez** [comment]
> > Oh @ArnoldSmith86 this is actually a problem....
> > 
> > We now are limiting the size and position to be equal in both coodinates
> 
> Just override `handleCSS`.

Took me a while to understand this. Position still seems a problem.

**Comment by 96LawDawg** [comment]
Position does not allow full customization at a precise level, but it is much more intuitive than all that phX and phY stuff.  Seriously, 99% of the users just want to change the color and maybe move it to one edge or the other.  I can't imagine there is much need to be pixel precise on the piles.  I know we have the capability to do that, but is it necessary?  I say what is currently in this PR is more than enough.

**Comment by ArnoldSmith86** [comment]
I don't see how this does not give you all the flexibility you would ever want: http://212.47.248.129:3905/469a

**Comment by rogerl50** [comment]
#646 references this PR and suggests that they be merged together. Just putting a note here referencing *that* PR.

**Comment by ArnoldSmith86** [comment]
I think we should merge them at roughly the same time but I don't see a reason to merge one into the other. They are both about piles but technically completely unrelated.

**Comment by 96LawDawg** [comment]
Other than #646, this isn't waiting on anything, right?  Let's merge it.

**Pull request merged**
